### ISSUE TITLE
[Feature Request] DHooks x64

### ISSUE THREAD

**Issue by sapphonie** [alliedmodders/sourcemod#2114]
With the looming spectre of TF2's x64 branch, there hasn't been much discussion on updating DHooks to support x64, which a significant amount of plugins rely on. I figured I would make an issue so there is a place to keep track of discussion about that.

**Comment by peace-maker** [comment]
DHooks is built on two libraries to allow dynamic hooking of functions during runtime which both only support x86.

[SoureHook AutoGen](https://github.com/alliedmodders/metamod-source/blob/37408219baae58539c7298e9a30227bb65ae3679/core/sourcehook/sourcehook_hookmangen.cpp) to hijack virtual functions with arbitrary signatures
[DynamicHooks](https://github.com/Ayuto/DynamicHooks) (modified) for non-virtual functions

Both need individual support for x86_64 codegen and the respective calling convention, while technically all functions can be detoured, so for basic functionality DynamicHooks could be enough to upgrade first.
The [macro assembler library](https://github.com/alliedmodders/sourcepawn/tree/50d0675528f1ee5f633a1324f8ddcf7d233bd9cf/vm/x64) in the sourcepawn vm might need some additional instructions too.

**Comment by bottiger1** [comment]
Is anyone working on this?

Dynamichooks appears to have been forked for 64bit support but it requires C++20 and 4 more libs so I don't know if that would be accepted into sourcemod. https://github.com/qubka/DynoHook

**Comment by dvander** [comment]
What libs? As long as no coroutines cxx20 might be fineOn May 7, 2024 12:32 PM, bottiger1 ***@***.***> wrote:
Is anyone working on this?
Dynamichooks appears to have been forked for 64bit support but it requires C++20 and 4 more libs so I don't know if that would be accepted into sourcemod. https://github.com/qubka/DynoHook

**Comment by bottiger1** [comment]
asmjit asmtk catch2 zydis

Note that I'm just pointing out that there's a replacement for dynamichooks, I'm not volunteering to update dhooks. 

I took a quick glance at the code and it looks difficult to get familiarized with and I only have maybe 5 sourcepawn plugins that use it only because I was too lazy to make an extension.

**Comment by Kenzzer** [comment]
All x64 games

**Comment by Niterux** [comment]
So it's basically over for DHooks on Counter-Strike: Source?

**Comment by Kenzzer** [comment]
No, you can run the 32 bits version of the dedicated server and continue using dhooks. But yes dhooks x64 support is on going, and won't be finalised for a while

**Comment by yurakone** [comment]
since valve dropped 32 bit support in ssdk 2013 i was faced with a dilemma, use the stock server binaries(x86) with dhooks plugins , or use custom server binaries(x64) without plugins..

**Comment by Adrianilloo** [comment]
> Would love to see progress on this since the newly released TF2C only has 64-bit binaries

I'm interested as well. Especially for Linux.
